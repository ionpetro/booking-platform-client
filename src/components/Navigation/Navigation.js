import AuthService from '../../shared/services/auth.service';

const authService = new AuthService();

export default {
  name: 'Navigation',
  data: () => ({
    user: null,
    showDropDown: false,
    scrollPosition: undefined,
    openMobileNav: false
  }),
  destroyed() {
    window.removeEventListener('scroll', this.updateScroll);
  },
  mounted() {
    window.addEventListener('scroll', this.updateScroll);
    this.user =
      JSON.parse(localStorage.getItem('user')) || JSON.parse(sessionStorage.getItem('user'));
  },
  methods: {
    updateScroll() {
      this.scrollPosition = window.scrollY;
    },
    toggleDropdown() {
      this.showDropDown = !this.showDropDown;
    },
    closeDropdown() {
      this.showDropDown = false;
    },
    toggleMobileNav() {
      this.openMobileNav = !this.openMobileNav;
    },
    logOut() {
      authService.logout();
      this.$router.push('/login');
    }
  }
};
