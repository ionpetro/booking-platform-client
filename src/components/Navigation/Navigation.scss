.nav {
  background-color: $white;
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 22px 32px;
  position: sticky;
  top: 0;
  z-index: 100;
  transition: box-shadow 0.1s ease-in;

  &__scrolled {
    box-shadow: 0 6px 8px -6px $gray-tint;
  }

  a {
    color: $black;
    text-decoration: none;
    &:focus {
      outline: none;
    }

    &:focus-visible {
      outline: none;
      box-shadow: 0 0 0 2px $blue-light;
    }
    @media #{$mobile} {
      color: $white;
    }
  }

  &__logo {
    width: 116px;
    height: 20px;
    @media #{$mobile} {
      width: 86px;
      height: 15px;
    }
  }

  &__mobile {
    display: none;

    @media #{$mobile} {
      display: block;
    }
  }

  &__icon {
    display: none;

    .bar {
      transition: all 0.3s ease-in-out;
      width: 16px;
      border: 1.2px solid $secondary;
      border-radius: 10px;
      background-color: $secondary;
    }

    .bar:nth-child(2) {
      margin: 4px auto;
    }

    @media #{$mobile} {
      display: grid;
    }
  }

  &__links {
    display: flex;
    align-items: center;
    flex-direction: row;

    &__link {
      margin: 0 20px;

      &--white {
        @media #{$mobile} {
          color: $white;
        }
      }

      @media #{$mobile} {
        padding: 20px;
      }
    }

    &__line {
      @media #{$mobile} {
        margin-top: 100px;
        color: $white;
        background-color: $white;
        width: 30%;
      }
    }

    @media #{$mobile} {
      z-index: 100;
      flex-direction: column;
      position: fixed;
      justify-content: center;
      top: 55px;
      background-color: $secondary;
      width: 100%;
      height: calc(100vh - 55px);
      transform: translate(calc(-100% - 16px));
      transition: transform 0.7s ease-in-out;
      text-align: center;
      overflow: hidden;
    }
  }

  &__mobile-nav {
    @media #{$mobile} {
      transform: translate(calc(0% - 16px));
    }
  }

  &__login {
    &__dropdown {
      &__title {
        cursor: pointer;
        display: flex;
        &__name {
          padding-right: 4px;
        }

        &__arrow {
          &__reverse {
            transition: transform 0.3s ease-in-out;
            transform: rotate(180deg);
          }

          &__reset {
            transition: transform 0.3s ease-in-out;
            transform: rotate(0deg);
          }
        }
      }

      &__body {
        position: absolute;
        width: 130px;
        box-shadow: 0px 4px 15px rgba(0, 0, 0, 0.25);
        border-radius: 4px;
        height: 48px;
        top: 48px + 10px;
        right: 30px;
        background-color: $secondary;
        display: flex;
        justify-content: center;
        align-items: center;
        cursor: pointer;

        &:hover {
          span {
            text-decoration: underline;
          }
        }

        span {
          padding-left: 10px;
          color: $white;
        }
      }
    }
    @media #{$mobile} {
      display: none;
    }
  }

  &__icon.active .bar:nth-child(2) {
    opacity: 0;
  }

  &__icon.active .bar:nth-child(1) {
    transform: translateY(6px) rotate(45deg);
  }

  &__icon.active .bar:nth-child(3) {
    transform: translateY(-6px) rotate(-45deg);
  }

  @media #{$mobile} {
    padding: 22px 16px;
    background-color: $white;
    margin: 0;
  }
}

// class for active routes
.is-active {
  font-weight: 700;
}
