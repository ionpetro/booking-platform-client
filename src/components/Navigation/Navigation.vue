<template>
  <nav :class="['nav', 'body--large', scrollPosition ? 'nav__scrolled' : '']">
    <img class="nav__logo" src="@/assets/images/blue.png" alt="LogoBlue" />
    <div @click="toggleMobileNav" :class="['nav__icon', openMobileNav ? 'active' : '']">
      <span class="bar"></span>
      <span class="bar"></span>
      <span class="bar"></span>
    </div>
    <div :class="['nav__links', openMobileNav ? 'nav__mobile-nav' : '']">
      <div v-if="user" class="nav__links__link nav__mobile nav__links__link--white">
        <strong>Hello, {{ user.firstName }}</strong>
      </div>
      <div class="nav__links__link">
        <router-link @click.native="toggleMobileNav" to="/">Home</router-link>
      </div>
      <div class="nav__links__link">
        <router-link @click.native="toggleMobileNav" to="/units">Units</router-link>
      </div>
      <hr class="nav__mobile nav__links__line" />
      <div v-if="user" class="nav__links__link nav__mobile"><a @click="logOut">Log out</a></div>
      <div v-else class="nav__links__link nav__mobile">
        <router-link v-if="!user" to="/login">Login</router-link>
      </div>
    </div>
    <div class="nav__login" v-click-outside="closeDropdown">
      <router-link v-if="!user" to="/login">Login</router-link>
      <div v-else class="nav__login__dropdown">
        <div class="nav__login__dropdown__title" @click.prevent="toggleDropdown">
          <span class="nav__login__dropdown__title__name" to="/units">{{ user.firstName }}</span>
          <img
            :class="
              showDropDown
                ? 'nav__login__dropdown__title__arrow__reverse'
                : 'nav__login__dropdown__title__arrow__reset'
            "
            src="@/assets/icons/Dropdown.svg"
            alt="arrow"
          />
        </div>
        <transition name="fade">
          <div v-if="showDropDown" class="nav__login__dropdown__body" @click="logOut">
            <img src="@/assets/icons/Logout.svg" alt="Logout" />
            <span>Logout</span>
          </div>
        </transition>
      </div>
    </div>
  </nav>
</template>

<script src="./Navigation.js"></script>

<style lang="scss" src="./Navigation.scss" scoped></style>
