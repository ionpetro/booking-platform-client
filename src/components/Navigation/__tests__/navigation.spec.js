import {shallowMount, createLocalVue} from '@vue/test-utils';
import flushPromises from 'flush-promises';
import Navigation from '../Navigation.vue';
import AuthService from '../../../shared/services/auth.service';

import clickOutsideDirective from '../../../shared/helpers/directives/clickOutside';

describe('Navigation', () => {
  let wrapper;

  const logoutMock = jest.fn();
  AuthService.prototype.logout = logoutMock;

  const $router = {push: jest.fn()};

  beforeEach(() => {
    clickOutsideDirective();

    wrapper = shallowMount(Navigation, {
      stubs: ['router-link'],
      mocks: {$router}
    });
    // mount component and login a user
    wrapper.setData({user: {username: 'test', firstName: 'test', lastName: 'test'}});
  });

  describe('when loaded', () => {
    it('renders correctly', () => {
      expect(wrapper.element).toMatchSnapshot();
    });
    it('opens dropdown', async () => {
      await wrapper.find('.nav__login__dropdown__title').trigger('click');

      await flushPromises();

      expect(wrapper.vm.showDropDown).toBeTruthy();
    });

    it('opens navbar menu on mobile', async () => {
      await wrapper.find('.nav__icon').trigger('click');
      await flushPromises();

      expect(wrapper.vm.openMobileNav).toBeTruthy();
    });

    it('logs user out when logout is clicked', async () => {
      // open dropdown
      await wrapper.setData({showDropDown: true});

      console.log(wrapper.vm.showDropDown);

      // close dropdown on click out

      await wrapper.find('.nav__login__dropdown__body').trigger('click');
      expect(AuthService.prototype.logout).toHaveBeenCalledTimes(1);
      expect($router.push).toHaveBeenCalledWith('/login');
    });

    it('adds class on scroll', async () => {
      const localVue = createLocalVue();
      wrapper = shallowMount(Navigation, {
        localVue,
        attachToDocument: true
      });

      await window.dispatchEvent(new CustomEvent('scroll'));
      expect(wrapper.vm.scrollPosition).toEqual(0);
    });
  });
});
