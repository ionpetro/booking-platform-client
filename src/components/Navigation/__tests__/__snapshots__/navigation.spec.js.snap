// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation when loaded renders correctly 1`] = `
<nav
  class="nav body--large"
>
  <img
    alt="LogoBlue"
    class="nav__logo"
    src="@/assets/images/blue.png"
  />
   
  <div
    class="nav__icon"
  >
    <span
      class="bar"
    />
     
    <span
      class="bar"
    />
     
    <span
      class="bar"
    />
  </div>
   
  <div
    class="nav__links"
  >
    <div
      class="nav__links__link nav__mobile nav__links__link--white"
    >
      <strong>
        Hello, test
      </strong>
    </div>
     
    <div
      class="nav__links__link"
    >
      <router-link-stub
        to="/"
      >
        About
      </router-link-stub>
    </div>
     
    <div
      class="nav__links__link"
    >
      <router-link-stub
        to="/units"
      >
        Units
      </router-link-stub>
    </div>
     
    <hr
      class="nav__mobile nav__links__line"
    />
     
    <div
      class="nav__links__link nav__mobile"
    >
      <a>
        Log out
      </a>
    </div>
  </div>
   
  <div
    class="nav__login"
  >
    <div
      class="nav__login__dropdown"
    >
      <div
        class="nav__login__dropdown__title"
      >
        <span
          class="nav__login__dropdown__title__name"
          to="/units"
        >
          test
        </span>
         
        <img
          alt="arrow"
          class="nav__login__dropdown__title__arrow__reset"
          src="@/assets/icons/Dropdown.svg"
        />
      </div>
       
      <transition-stub
        name="fade"
      >
        <!---->
      </transition-stub>
    </div>
  </div>
</nav>
`;
